File: device_gate/store.py
"""JSON-backed device record store.

One file (``devices.json``) holds every paired device. Records carry
the issued token, the device's tier, last-seen timestamp, and the
metadata the mobile app supplied at pairing time. The file is written
atomically (temp + rename) so a crash mid-write doesn't corrupt the
store.

Surface mirrors what the public API in :mod:`core` needs — the
package never exposes raw ``Device`` objects to GUI / Gateway callers;
both flows go through ``core.list_devices`` etc. which return dicts.
"""

from __future__ import annotations

import json
import os
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

from Core.shared.secure_file import harden_perms


# ── Tiers ──────────────────────────────────────────────────────────────


TIER_READ_ONLY = "read_only"
TIER_TOOL_USE = "tool_use"
TIER_DESTRUCTIVE = "destructive_tool_access"

ALL_TIERS = (TIER_READ_ONLY, TIER_TOOL_USE, TIER_DESTRUCTIVE)

# Each tier strictly contains the previous one. Numeric ranks make
# "is tier X >= tier Y" comparisons explicit at the call site.
TIER_RANK: Dict[str, int] = {
    TIER_READ_ONLY: 0,
    TIER_TOOL_USE: 1,
    TIER_DESTRUCTIVE: 2,
}


def is_valid_tier(tier: Any) -> bool:
    return isinstance(tier, str) and tier in ALL_TIERS


def tier_rank(tier: str) -> int:
    return TIER_RANK.get(tier, -1)


# ── Device record ──────────────────────────────────────────────────────


@dataclass
class Device:
    device_id: str
    name: str
    token: str
    tier: str = TIER_READ_ONLY
    paired_at: float = 0.0
    last_seen: float = 0.0
    metadata: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self, *, include_token: bool = False) -> Dict[str, Any]:
        out: Dict[str, Any] = {
            "device_id": self.device_id,
            "name": self.name,
            "tier": self.tier,
            "paired_at": float(self.paired_at),
            "last_seen": float(self.last_seen),
            "metadata": dict(self.metadata),
        }
        if include_token:
            out["token"] = self.token
        return out

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Device":
        return cls(
            device_id=str(d.get("device_id", "")),
            name=str(d.get("name", "")),
            token=str(d.get("token", "")),
            tier=str(d.get("tier", TIER_READ_ONLY) or TIER_READ_ONLY),
            paired_at=float(d.get("paired_at", 0.0) or 0.0),
            last_seen=float(d.get("last_seen", 0.0) or 0.0),
            metadata=dict(d.get("metadata", {}) or {}),
        )


# ── Store (JSON-backed) ────────────────────────────────────────────────


class DeviceStore:
    """Thread-safe device-record store. One JSON file on disk."""

    def __init__(self, path: Path) -> None:
        self.path = Path(path)
        self._lock = threading.RLock()

    def _load(self) -> List[Device]:
        if not self.path.exists():
            return []
        try:
            raw = json.loads(self.path.read_text(encoding="utf-8"))
        except Exception:  # noqa: BLE001
            return []
        items = raw.get("devices") if isinstance(raw, dict) else raw
        if not isinstance(items, list):
            return []
        return [Device.from_dict(it) for it in items if isinstance(it, dict)]

    def _save(self, devices: List[Device]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = {"devices": [d.to_dict(include_token=True) for d in devices]}
        tmp = self.path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        os.replace(tmp, self.path)
        # Live device bearer tokens — restrict to owner-only access.
        harden_perms(self.path)

    # ── Read-side ─────────────────────────────────────────────────────

    def list(self) -> List[Device]:
        with self._lock:
            return self._load()

    def get(self, device_id: str) -> Optional[Device]:
        with self._lock:
            for d in self._load():
                if d.device_id == device_id:
                    return d
        return None

    def find_by_token(self, token: str) -> Optional[Device]:
        if not isinstance(token, str) or not token:
            return None
        with self._lock:
            for d in self._load():
                if d.token == token:
                    return d
        return None

    # ── Write-side ────────────────────────────────────────────────────

    def add(self, device: Device) -> Device:
        with self._lock:
            devices = self._load()
            # Reject if device_id collision; caller should mint a fresh id.
            if any(d.device_id == device.device_id for d in devices):
                raise ValueError(f"device_id {device.device_id!r} already exists")
            devices.append(device)
            self._save(devices)
            return device

    def update(self, device_id: str, **fields: Any) -> Optional[Device]:
        with self._lock:
            devices = self._load()
            target = next((d for d in devices if d.device_id == device_id), None)
            if target is None:
                return None
            for k, v in fields.items():
                if hasattr(target, k):
                    setattr(target, k, v)
            self._save(devices)
            return target

    def remove(self, device_id: str) -> bool:
        with self._lock:
            devices = self._load()
            before = len(devices)
            devices = [d for d in devices if d.device_id != device_id]
            if len(devices) == before:
                return False
            self._save(devices)
            return True

    def touch(self, device_id: str, when: Optional[float] = None) -> None:
        """Update ``last_seen`` for a device. Idempotent; missing
        device_id is a no-op so a stale token check doesn't crash."""
        ts = float(when if when is not None else time.time())
        with self._lock:
            devices = self._load()
            changed = False
            for d in devices:
                if d.device_id == device_id:
                    d.last_seen = ts
                    changed = True
                    break
            if changed:
                self._save(devices)


# ── Action log (per-device rolling history) ────────────────────────────


# How many action entries we retain per device. The GUI only renders the
# most-recent handful; the cap keeps the JSON file bounded even for a
# device that's been rotated / re-tiered many times.
ACTION_LOG_CAP = 50


class ActionLog:
    """JSON-backed rolling log of GUI-driven mutations, keyed by device.

    Separate file from ``devices.json`` so the audit trail survives a
    device being revoked (the device row is gone, but the operator may
    still want to see "this device was revoked at T"). Same atomic
    temp+rename write discipline as :class:`DeviceStore`; the entries
    carry no secrets, so no ``harden_perms`` call here.

    Entry shape: ``{action, timestamp, status}`` where ``timestamp`` is
    ISO-8601 UTC. Stored oldest-first on disk; :meth:`recent` returns
    newest-first to match the GUI's display order.
    """

    def __init__(self, path: Path) -> None:
        self.path = Path(path)
        self._lock = threading.RLock()

    def _load(self) -> Dict[str, List[Dict[str, Any]]]:
        if not self.path.exists():
            return {}
        try:
            raw = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            return {}
        if not isinstance(raw, dict):
            return {}
        out: Dict[str, List[Dict[str, Any]]] = {}
        for did, entries in raw.items():
            if isinstance(did, str) and isinstance(entries, list):
                out[did] = [e for e in entries if isinstance(e, dict)]
        return out

    def _save(self, data: Dict[str, List[Dict[str, Any]]]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self.path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(data, indent=2), encoding="utf-8")
        os.replace(tmp, self.path)

    def record(
        self,
        device_id: str,
        action: str,
        *,
        status: str = "ok",
        timestamp: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Append one ``{action, timestamp, status}`` entry for a device.

        ``timestamp`` defaults to ``now()`` in ISO-8601 UTC. Oldest
        entries are dropped once the per-device list exceeds
        :data:`ACTION_LOG_CAP`. Returns the entry written so callers can
        assert on it in tests."""
        ts = timestamp or _utc_now_iso()
        entry = {"action": action, "timestamp": ts, "status": status}
        with self._lock:
            data = self._load()
            entries = data.setdefault(device_id, [])
            entries.append(entry)
            # Trim from the front — oldest-first on disk.
            if len(entries) > ACTION_LOG_CAP:
                del entries[: len(entries) - ACTION_LOG_CAP]
            self._save(data)
        return entry

    def recent(self, device_id: str, *, limit: int = 20) -> List[Dict[str, Any]]:
        """Return up to ``limit`` entries for ``device_id``, newest-first.

        Unknown device → empty list. ``limit`` is clamped to ``>= 0``."""
        limit = max(0, int(limit))
        with self._lock:
            entries = list(self._load().get(device_id, []))
        # Disk is oldest-first; reverse for newest-first, then cap.
        return list(reversed(entries))[:limit]


def _utc_now_iso() -> str:
    """Current UTC time as a second-resolution ISO-8601 string with a
    trailing ``Z`` (e.g. ``2026-05-30T12:34:56Z``) — matches the format
    the GUI's relative-time parser expects."""
    from datetime import datetime, timezone

    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


__all__ = [
    "TIER_READ_ONLY",
    "TIER_TOOL_USE",
    "TIER_DESTRUCTIVE",
    "ALL_TIERS",
    "TIER_RANK",
    "is_valid_tier",
    "tier_rank",
    "Device",
    "DeviceStore",
    "ACTION_LOG_CAP",
    "ActionLog",
]

File: rust/crates/wylde-device-gate/src/store.rs
//! JSON-backed device record store + tier constants.
//!
//! Rust port of `device_gate/store.py`. One file (`devices.json`) holds every
//! paired device; writes are atomic (tmpfile + rename) so a crash mid-write
//! doesn't corrupt the store. The Python and Rust implementations share the
//! same file format, so a live cutover via the strangler-fig flag doesn't
//! invalidate existing pairings.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use serde_json::Value;

// ── Tiers ──────────────────────────────────────────────────────────────

/// Default tier assigned at pairing time. View-only access; cannot invoke tools.
pub const TIER_READ_ONLY: &str = "read_only";
/// Mid tier. Non-destructive tool calls (read / search / retrieve).
pub const TIER_TOOL_USE: &str = "tool_use";
/// Full surface, including write / delete / execute.
pub const TIER_DESTRUCTIVE: &str = "destructive_tool_access";

/// Tier list in rank order. Matches Python's `ALL_TIERS`.
pub const ALL_TIERS: [&str; 3] = [TIER_READ_ONLY, TIER_TOOL_USE, TIER_DESTRUCTIVE];

/// Numeric rank for "is tier X >= tier Y" comparisons at the call site.
pub fn tier_rank(tier: &str) -> i32 {
    match tier {
        TIER_READ_ONLY => 0,
        TIER_TOOL_USE => 1,
        TIER_DESTRUCTIVE => 2,
        _ => -1,
    }
}

pub fn is_valid_tier(tier: &str) -> bool {
    ALL_TIERS.contains(&tier)
}

// ── Device record ──────────────────────────────────────────────────────

/// One paired device. Wire shape (the JSON written into `devices.json`)
/// matches Python's `Device.to_dict(include_token=True)`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Device {
    pub device_id: String,
    pub name: String,
    pub token: String,
    #[serde(default = "default_tier")]
    pub tier: String,
    #[serde(default)]
    pub paired_at: f64,
    #[serde(default)]
    pub last_seen: f64,
    #[serde(default)]
    pub metadata: HashMap<String, Value>,
}

fn default_tier() -> String {
    TIER_READ_ONLY.to_string()
}

impl Device {
    /// GUI-facing dict — drops the token field. Mirrors Python's
    /// `Device.to_dict(include_token=False)`.
    pub fn to_public_value(&self) -> Value {
        serde_json::json!({
            "device_id": self.device_id,
            "name": self.name,
            "tier": self.tier,
            "paired_at": self.paired_at,
            "last_seen": self.last_seen,
            "metadata": self.metadata,
        })
    }
}

// ── Store ──────────────────────────────────────────────────────────────

/// On-disk envelope: `{"devices": [Device, ...]}`. Kept as its own struct
/// so the JSON shape matches Python exactly.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
struct DeviceFile {
    #[serde(default)]
    devices: Vec<Device>,
}

/// Thread-safe device-record store. One JSON file on disk.
///
/// Each operation re-reads the file (matching the Python store's behaviour
/// where the in-memory cache is the file itself). This keeps multi-process
/// safety simple: the only shared state is the file, and writes are atomic.
pub struct DeviceStore {
    path: PathBuf,
    lock: Mutex<()>,
}

impl DeviceStore {
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self {
            path: path.into(),
            lock: Mutex::new(()),
        }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    fn load(&self) -> Vec<Device> {
        if !self.path.exists() {
            return Vec::new();
        }
        let raw = match std::fs::read_to_string(&self.path) {
            Ok(s) => s,
            Err(_) => return Vec::new(),
        };
        // Accept both wrapped (`{"devices": [...]}`) and bare-array shapes so
        // any historic file layout still loads. Python's loader has the same
        // bivalent reader.
        if let Ok(file) = serde_json::from_str::<DeviceFile>(&raw) {
            return file.devices;
        }
        if let Ok(arr) = serde_json::from_str::<Vec<Device>>(&raw) {
            return arr;
        }
        Vec::new()
    }

    fn save(&self, devices: &[Device]) -> std::io::Result<()> {
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let file = DeviceFile {
            devices: devices.to_vec(),
        };
        let body = serde_json::to_string_pretty(&file).map_err(std::io::Error::other)?;
        let tmp = self.path.with_extension("json.tmp");
        std::fs::write(&tmp, body.as_bytes())?;
        std::fs::rename(&tmp, &self.path)?;
        // Live device bearer tokens — restrict to owner-only access.
        // Fail-soft: a hardening failure must not fail the write.
        let _ = wylde_shared::secure_file::harden_perms(&self.path);
        Ok(())
    }

    // ── Read-side ─────────────────────────────────────────────────────

    pub fn list(&self) -> Vec<Device> {
        let _g = self.lock.lock().expect("device store poisoned");
        self.load()
    }

    pub fn get(&self, device_id: &str) -> Option<Device> {
        let _g = self.lock.lock().expect("device store poisoned");
        self.load().into_iter().find(|d| d.device_id == device_id)
    }

    pub fn find_by_token(&self, token: &str) -> Option<Device> {
        if token.is_empty() {
            return None;
        }
        let _g = self.lock.lock().expect("device store poisoned");
        self.load().into_iter().find(|d| d.token == token)
    }

    // ── Write-side ────────────────────────────────────────────────────

    /// Append a fresh device record. Returns an error if `device_id`
    /// already exists — the caller is expected to mint a unique id.
    pub fn add(&self, device: Device) -> Result<Device, StoreError> {
        let _g = self.lock.lock().expect("device store poisoned");
        let mut devices = self.load();
        if devices.iter().any(|d| d.device_id == device.device_id) {
            return Err(StoreError::Duplicate(device.device_id));
        }
        devices.push(device.clone());
        self.save(&devices).map_err(StoreError::Io)?;
        Ok(device)
    }

    /// Apply `updater` to the matching record. Returns the updated copy
    /// or `None` if the device wasn't found. Mirrors Python's `update`,
    /// which took `**fields` kwargs; the closure form is the Rust idiom.
    pub fn update<F>(&self, device_id: &str, updater: F) -> Option<Device>
    where
        F: FnOnce(&mut Device),
    {
        let _g = self.lock.lock().expect("device store poisoned");
        let mut devices = self.load();
        let target = devices.iter_mut().find(|d| d.device_id == device_id)?;
        updater(target);
        let updated = target.clone();
        let _ = self.save(&devices);
        Some(updated)
    }

    pub fn remove(&self, device_id: &str) -> bool {
        let _g = self.lock.lock().expect("device store poisoned");
        let mut devices = self.load();
        let before = devices.len();
        devices.retain(|d| d.device_id != device_id);
        if devices.len() == before {
            return false;
        }
        let _ = self.save(&devices);
        true
    }

    /// Update `last_seen`. Missing device_id is a no-op so a stale token
    /// check doesn't crash. Matches Python's `touch`.
    pub fn touch(&self, device_id: &str, when: f64) {
        let _g = self.lock.lock().expect("device store poisoned");
        let mut devices = self.load();
        let mut changed = false;
        for d in &mut devices {
            if d.device_id == device_id {
                d.last_seen = when;
                changed = true;
                break;
            }
        }
        if changed {
            let _ = self.save(&devices);
        }
    }
}

/// Errors returned by [`DeviceStore::add`].
#[derive(Debug, thiserror::Error)]
pub enum StoreError {
    #[error("device_id {0:?} already exists")]
    Duplicate(String),
    #[error("device store IO: {0}")]
    Io(#[from] std::io::Error),
}

// ── Action log (per-device rolling history) ────────────────────────────

/// How many action entries we retain per device. Matches Python's
/// `store.ACTION_LOG_CAP`. The GUI only renders the most-recent handful;
/// the cap keeps the JSON file bounded even for a device that's been
/// rotated / re-tiered many times.
pub const ACTION_LOG_CAP: usize = 50;

/// One audit entry. Wire shape matches Python's `ActionLog` entry —
/// `{action, timestamp, status}` where `timestamp` is ISO-8601 UTC,
/// second resolution (e.g. `2026-05-30T12:34:56Z`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionEntry {
    pub action: String,
    pub timestamp: String,
    #[serde(default = "default_status")]
    pub status: String,
}

fn default_status() -> String {
    "ok".to_string()
}

/// JSON-backed rolling log of GUI-driven mutations, keyed by device.
///
/// Rust port of `device_gate/store.py::ActionLog`. Separate file from
/// `devices.json` so the audit trail survives a device being revoked (the
/// device row is gone, but the operator may still want to see "this device
/// was revoked at T"). Same atomic temp+rename write discipline as
/// [`DeviceStore`]; the entries carry no secrets, so no `harden_perms` here.
///
/// Stored oldest-first on disk (the JSON map is `{device_id: [entry, ...]}`);
/// [`ActionLog::recent`] returns newest-first to match the GUI's display
/// order.
pub struct ActionLog {
    path: PathBuf,
    lock: Mutex<()>,
}

impl ActionLog {
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self {
            path: path.into(),
            lock: Mutex::new(()),
        }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    fn load(&self) -> HashMap<String, Vec<ActionEntry>> {
        if !self.path.exists() {
            return HashMap::new();
        }
        let raw = match std::fs::read_to_string(&self.path) {
            Ok(s) => s,
            Err(_) => return HashMap::new(),
        };
        // Mirror Python's lenient loader: any parse error → empty map rather
        // than a crash, so a corrupt file degrades to "no history" instead of
        // taking the service down.
        serde_json::from_str::<HashMap<String, Vec<ActionEntry>>>(&raw).unwrap_or_default()
    }

    fn save(&self, data: &HashMap<String, Vec<ActionEntry>>) -> std::io::Result<()> {
        if let Some(parent) = self.path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let body = serde_json::to_string_pretty(data).map_err(std::io::Error::other)?;
        let tmp = self.path.with_extension("json.tmp");
        std::fs::write(&tmp, body.as_bytes())?;
        std::fs::rename(&tmp, &self.path)?;
        Ok(())
    }

    /// Append one `{action, timestamp, status}` entry for a device.
    /// `timestamp` is `now()` in ISO-8601 UTC; oldest entries are dropped
    /// once the per-device list exceeds [`ACTION_LOG_CAP`]. Returns the entry
    /// written so callers can assert on it in tests. Matches Python's
    /// `ActionLog.record`.
    pub fn record(&self, device_id: &str, action: &str, status: &str) -> ActionEntry {
        self.record_at(device_id, action, status, &utc_now_iso())
    }

    /// `record` with an explicit timestamp — the test seam Python exposes via
    /// its `timestamp=` kwarg.
    pub fn record_at(
        &self,
        device_id: &str,
        action: &str,
        status: &str,
        timestamp: &str,
    ) -> ActionEntry {
        let entry = ActionEntry {
            action: action.to_string(),
            timestamp: timestamp.to_string(),
            status: status.to_string(),
        };
        let _g = self.lock.lock().expect("action log poisoned");
        let mut data = self.load();
        let entries = data.entry(device_id.to_string()).or_default();
        entries.push(entry.clone());
        // Trim from the front — oldest-first on disk.
        if entries.len() > ACTION_LOG_CAP {
            let overflow = entries.len() - ACTION_LOG_CAP;
            entries.drain(0..overflow);
        }
        let _ = self.save(&data);
        entry
    }

    /// Return up to `limit` entries for `device_id`, newest-first. Unknown
    /// device → empty list. Matches Python's `ActionLog.recent`.
    pub fn recent(&self, device_id: &str, limit: usize) -> Vec<ActionEntry> {
        let _g = self.lock.lock().expect("action log poisoned");
        let data = self.load();
        let Some(entries) = data.get(device_id) else {
            return Vec::new();
        };
        // Disk is oldest-first; reverse for newest-first, then cap.
        entries.iter().rev().take(limit).cloned().collect()
    }
}

/// Current UTC time as a second-resolution ISO-8601 string with a trailing
/// `Z` (e.g. `2026-05-30T12:34:56Z`) — matches Python's `_utc_now_iso` and
/// the format the GUI's relative-time parser expects.
fn utc_now_iso() -> String {
    chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn fresh_store() -> (TempDir, DeviceStore) {
        let tmp = TempDir::new().expect("tempdir");
        let path = tmp.path().join("devices.json");
        (tmp, DeviceStore::new(path))
    }

    fn make_device(id: &str, token: &str) -> Device {
        Device {
            device_id: id.into(),
            name: format!("name-{id}"),
            token: token.into(),
            tier: TIER_READ_ONLY.into(),
            paired_at: 1_000_000.0,
            last_seen: 1_000_000.0,
            metadata: HashMap::new(),
        }
    }

    #[test]
    fn add_get_remove_roundtrip() {
        let (_tmp, store) = fresh_store();
        store.add(make_device("d1", "t1")).unwrap();
        assert!(store.get("d1").is_some());
        assert_eq!(store.find_by_token("t1").unwrap().device_id, "d1");
        assert!(store.remove("d1"));
        assert!(store.get("d1").is_none());
        assert!(store.find_by_token("t1").is_none());
    }

    #[test]
    fn add_rejects_duplicate_device_id() {
        let (_tmp, store) = fresh_store();
        store.add(make_device("d1", "t1")).unwrap();
        let err = store.add(make_device("d1", "t2"));
        assert!(matches!(err, Err(StoreError::Duplicate(_))));
    }

    #[test]
    fn update_changes_fields() {
        let (_tmp, store) = fresh_store();
        store.add(make_device("d1", "t1")).unwrap();
        let updated = store
            .update("d1", |d| d.tier = TIER_TOOL_USE.into())
            .unwrap();
        assert_eq!(updated.tier, TIER_TOOL_USE);
        assert_eq!(store.get("d1").unwrap().tier, TIER_TOOL_USE);
    }

    #[test]
    fn update_returns_none_for_missing() {
        let (_tmp, store) = fresh_store();
        assert!(store
            .update("nope", |d| d.tier = TIER_TOOL_USE.into())
            .is_none());
    }

    #[test]
    fn touch_updates_last_seen() {
        let (_tmp, store) = fresh_store();
        store.add(make_device("d1", "t1")).unwrap();
        store.touch("d1", 2_000_000.0);
        assert_eq!(store.get("d1").unwrap().last_seen, 2_000_000.0);
    }

    #[test]
    fn touch_missing_is_noop() {
        let (_tmp, store) = fresh_store();
        store.touch("nope", 999.0); // must not panic.
    }

    #[test]
    fn missing_file_returns_empty_list() {
        let (_tmp, store) = fresh_store();
        assert!(store.list().is_empty());
    }

    #[test]
    fn parses_python_wrapped_format() {
        let (tmp, store) = fresh_store();
        let path = tmp.path().join("devices.json");
        std::fs::write(
            &path,
            r#"{"devices": [{"device_id": "x", "name": "n", "token": "t", "tier": "tool_use", "paired_at": 1.0, "last_seen": 2.0, "metadata": {}}]}"#,
        )
        .unwrap();
        let list = store.list();
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].tier, "tool_use");
    }

    fn fresh_action_log() -> (TempDir, ActionLog) {
        let tmp = TempDir::new().expect("tempdir");
        let path = tmp.path().join("action_log.json");
        (tmp, ActionLog::new(path))
    }

    #[test]
    fn action_log_records_and_returns_newest_first() {
        let (_tmp, log) = fresh_action_log();
        log.record_at("d1", "paired", "ok", "2026-05-30T12:00:00Z");
        log.record_at("d1", "tier → tool_use", "ok", "2026-05-30T12:01:00Z");
        log.record_at("d1", "revoked", "ok", "2026-05-30T12:02:00Z");
        let recent = log.recent("d1", 20);
        assert_eq!(recent.len(), 3);
        // Newest-first.
        assert_eq!(recent[0].action, "revoked");
        assert_eq!(recent[1].action, "tier → tool_use");
        assert_eq!(recent[2].action, "paired");
        assert_eq!(recent[0].status, "ok");
    }

    #[test]
    fn action_log_unknown_device_is_empty() {
        let (_tmp, log) = fresh_action_log();
        log.record("d1", "paired", "ok");
        assert!(log.recent("nope", 20).is_empty());
    }

    #[test]
    fn action_log_respects_limit() {
        let (_tmp, log) = fresh_action_log();
        for i in 0..10 {
            log.record_at("d1", &format!("a{i}"), "ok", "2026-05-30T12:00:00Z");
        }
        assert_eq!(log.recent("d1", 3).len(), 3);
    }

    #[test]
    fn action_log_caps_at_fifty_oldest_dropped() {
        let (_tmp, log) = fresh_action_log();
        for i in 0..(ACTION_LOG_CAP + 5) {
            log.record_at("d1", &format!("a{i}"), "ok", "2026-05-30T12:00:00Z");
        }
        let recent = log.recent("d1", 100);
        assert_eq!(recent.len(), ACTION_LOG_CAP);
        // Newest entry retained; the five oldest dropped from the front.
        assert_eq!(recent[0].action, format!("a{}", ACTION_LOG_CAP + 4));
        assert_eq!(recent[ACTION_LOG_CAP - 1].action, "a5");
    }

    #[test]
    fn action_log_survives_reload() {
        let tmp = TempDir::new().expect("tempdir");
        let path = tmp.path().join("action_log.json");
        ActionLog::new(&path).record_at("d1", "paired", "ok", "2026-05-30T12:00:00Z");
        // Fresh instance over the same file — persistence across restart.
        let recent = ActionLog::new(&path).recent("d1", 20);
        assert_eq!(recent.len(), 1);
        assert_eq!(recent[0].action, "paired");
    }

    #[test]
    fn tier_helpers_align_with_python() {
        assert_eq!(tier_rank(TIER_READ_ONLY), 0);
        assert_eq!(tier_rank(TIER_TOOL_USE), 1);
        assert_eq!(tier_rank(TIER_DESTRUCTIVE), 2);
        assert_eq!(tier_rank("nope"), -1);
        assert!(is_valid_tier(TIER_TOOL_USE));
        assert!(!is_valid_tier("superuser"));
    }
}
